client/jetpack-cloud/sections/partner-portal/credit-card-fields/credit-card-submit-button.tsx

```diff
@@ -0,0 +1,91 @@
+import { Button, FormStatus, useFormStatus } from '@automattic/composite-checkout';
+import { useElements, CardElement } from '@stripe/react-stripe-js';
+import { useSelect } from '@wordpress/data';
+import { useI18n } from '@wordpress/react-i18n';
+import debugFactory from 'debug';
+
+const debug = debugFactory( 'calypso:composite-checkout:credit-card' );
+
+export default function CreditCardSubmitButton( {
+	disabled,
+	onClick,
+	store,
+	stripe,
+	stripeConfiguration,
+	activeButtonText,
+} ) {
+	const { __ } = useI18n();
+	const fields = useSelect( ( select ) => select( 'credit-card' ).getFields() );
+	const useAsPrimaryPaymentMethod = useSelect( ( select ) =>
+		select( 'credit-card' ).useAsPrimaryPaymentMethod()
+	);
+	const cardholderName = fields.cardholderName;
+	const { formStatus } = useFormStatus();
+	const elements = useElements();
+	const cardElement = elements?.getElement( CardElement ) ?? undefined;
+
+	return (
+		<Button
+			// eslint-disable-next-line wpcalypso/jsx-classname-namespace
+			className="button is-primary"
+			disabled={ disabled }
+			onClick={ () => {
+				if ( isCreditCardFormValid( store, __ ) ) {
+					debug( 'submitting stripe payment' );
+					onClick( 'card', {
+						stripe,
+						name: cardholderName?.value,
+						stripeConfiguration,
+						cardElement,
+						useAsPrimaryPaymentMethod,
+					} );
+					return;
+				}
+			} }
+			buttonType="primary"
+			isBusy={ FormStatus.SUBMITTING === formStatus }
+			fullWidth
+		>
+			<ButtonContents formStatus={ formStatus } activeButtonText={ activeButtonText } />
+		</Button>
+	);
+}
+
+function ButtonContents( { formStatus, activeButtonText = undefined } ) {
+	const { __ } = useI18n();
+	if ( formStatus === FormStatus.SUBMITTING ) {
+		return __( 'Processing…' );
+	}
+	if ( formStatus === FormStatus.READY ) {
+		return activeButtonText || __( 'Save payment method' );
+	}
+	return __( 'Please wait…' );
+}
+
+function isCreditCardFormValid( store, __ ) {
```

Comment:
I feel this function is doing too much, and its name does not represent all that it does. Specifically, the function, on the one hand, computes whether the form is valid, and on the other, it dispatches actions to update the store.

One way to split the concerns could be to make the function only compute whether the form is valid and return an object that says whether it succeeded and, if not, the error that caused the form to be invalid. Then, we could dispatch whatever action is necessary according to the result from the call site instead of from the very same function.

I would also wrap the entire function declared in the `onClick` prop with `useCallback`.

Another thought is that sometimes there is nothing wrong with writing logic like the one present in the `isCreditCardFormValid` in the same component function. For example, I don't see anything wrong with something like:

```
export default function CreditCardSubmitButton( {
	disabled,
	onClick,
	store,
	stripe,
	stripeConfiguration,
	activeButtonText,
} ) {
	const { __ } = useI18n();
	const fields = useSelect( ( select ) => select( 'credit-card' ).getFields() );
	const useAsPrimaryPaymentMethod = useSelect( ( select ) =>
		select( 'credit-card' ).useAsPrimaryPaymentMethod()
	);
	const cardholderName = fields.cardholderName;
	const { formStatus } = useFormStatus();
	const elements = useElements();
	const cardElement = elements?.getElement( CardElement ) ?? undefined;

	// Previously, this next logic was contained in the isCreditCardFormValid function
	const onSubmit = useCallback( () => {
		const errors = store.selectors.getCardDataErrors( store.getState() );
		const incompleteFieldKeys = store.selectors.getIncompleteFieldKeys( store.getState() );
		const areThereErrors = Object.keys( errors ).some( ( errorKey ) => errors[ errorKey ] );
		const isValid =
			areThereErrors || ! cardholderName?.value.length || incompleteFieldKeys.length > 0;

		if ( ! cardholderName?.value.length ) {
			// Touch the field so it displays a validation error
			store.dispatch( store.actions.setFieldValue( 'cardholderName', '' ) );
			store.dispatch(
				store.actions.setFieldError( 'cardholderName', __( 'This field is required' ) )
			);
		}

		if ( incompleteFieldKeys.length > 0 ) {
			incompleteFieldKeys.map( ( key: string ) =>
				store.dispatch( store.actions.setCardDataError( key, __( 'This field is required' ) ) )
			);
		}
		if ( isValid ) {
			debug( 'submitting stripe payment' );
			onClick( 'card', {
				stripe,
				name: cardholderName?.value,
				stripeConfiguration,
				cardElement,
				useAsPrimaryPaymentMethod,
			} );
			return;
		}
	}, [
		stripe,
		onClick,
		useAsPrimaryPaymentMethod,
		cardholderName,
		cardElement,
		store,
		stripeConfiguration,
		__,
	] );

	return (
		<Button
			// eslint-disable-next-line wpcalypso/jsx-classname-namespace
			className="button is-primary"
			disabled={ disabled }
			onClick={ onSubmit }
			buttonType="primary"
			isBusy={ FormStatus.SUBMITTING === formStatus }
			fullWidth
		>
			<ButtonContents formStatus={ formStatus } activeButtonText={ activeButtonText } />
		</Button>
	);
}
```

Reply:
Thanks for suggesting this! I tried your code, but it didn't work (the form isn't submitting), could it be because we are actually using the [CheckoutSubmitButton](https://github.com/Automattic/wp-calypso/blob/165bf1cd39200e416bc329580e04b77a302480bc/packages/composite-checkout/README.md#checkoutsubmitbutton)?

However, the `buttonContents` [is now part of the CreditCardSubmitButton function](https://github.com/Automattic/wp-calypso/pull/61001#discussion_r805014885), which simplified things a bit.

Reply:
> could it be because we are actually using the [CheckoutSubmitButton](https://github.com/Automattic/wp-calypso/blob/165bf1cd39200e416bc329580e04b77a302480bc/packages/composite-checkout/README.md#checkoutsubmitbutton)?

Yes, that could have been the reason.

> However, the buttonContents https://github.com/Automattic/wp-calypso/pull/61001#discussion_r805014885, which simplified things a bit.

Great!